op-node/rollup/derive/engine_queue.go

```diff
@@ -399,6 +412,8 @@ func (eq *EngineQueue) postProcessSafeL2() {
 		if last.L2Block != eq.ec.SafeL2Head() { // avoid logging if there are no changes
 			last.L2Block = eq.ec.SafeL2Head()
 			eq.log.Debug("updated finality-data", "last_l1", last.L1Block, "last_l2", last.L2Block)
+			// Notify the new safe head
+			eq.safeHeadNotifs(eq.ec.SafeL2Head(), eq.origin.ID())
```

Comment:
The listener model seemed like a good idea to keep this out of the derivation pipeline, but if we fail to write the update to the DB we need to able to return an error and have the pipeline retry.  I suspect that's not safe at this point since various in-memory values will have already been updated. It may need to trigger a pipeline reset? Not returning an error isn't going to be an option though.

Reply:
Interestingly, the appropriate way to deal with failing to write is actually just to retry exactly the same write operation.  Even if we reset the pipeline it should just come back round to the same safe head update to be recorded. Actually getting reset to do that seems very difficult though because it doesn't have the L1 source of the current safe head.  I've got it to the point where it seems to reset correctly and start from the _next_ update but it still misses the update that failed to write.

I think the right answer is actually to just keep retrying the write in the safe head update listener until it succeeds. Getting a shutdown signal while retrying is still problematic though as the update would get lost if the node halted without writing it because the safe head in op-geth has already updated.